aggregator/handler/forward.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+)
+
+const (
+	tcpProtocol = "tcp"
+)
+
+var (
+	errEmptyServerList = errors.New("empty server list")
+	errHandlerClosed   = errors.New("handler is closed")
+	errBufferQueueFull = errors.New("buffer queue is full")
+)
+
+type dropType int
+
+const (
+	dropCurrent dropType = iota
+	dropOldestInQueue
+)
+
+type dialWithTimeoutFn func(protocol string, addr string, timeout time.Duration) (net.Conn, error)
+type writeToConnFn func(conn *net.TCPConn, data []byte) (int, error)
+
+type forwardHandler struct {
+	sync.RWMutex
+
+	servers             []string
+	log                 xlog.Logger
+	connectTimeout      time.Duration
+	connectionKeepAlive bool
+	reconnectRetrier    xretry.Retrier
+
+	bufCh             chan msgpack.Buffer
+	dialWithTimeoutFn dialWithTimeoutFn
+	writeToConnFn     writeToConnFn
+	wg                sync.WaitGroup
+	closed            bool
+}
+
+// NewForwardHandler creates a new forwarding handler.
+// TODO(xichen): add metrics.
+func NewForwardHandler(
+	servers []string,
+	opts ForwardHandlerOptions,
+) (aggregator.Handler, error) {
+	numServers := len(servers)
+	if numServers == 0 {
+		return nil, errEmptyServerList
+	}
+
+	h := &forwardHandler{
+		servers:             servers,
+		log:                 opts.InstrumentOptions().Logger(),
+		connectTimeout:      opts.ConnectTimeout(),
+		connectionKeepAlive: opts.ConnectionKeepAlive(),
+		reconnectRetrier:    opts.ReconnectRetrier(),
+		bufCh:               make(chan msgpack.Buffer, opts.QueueSize()),
+		dialWithTimeoutFn:   net.DialTimeout,
+		writeToConnFn:       writeToConn,
+	}
+
+	// Initiate the connections.
+	h.wg.Add(numServers)
+	for _, addr := range servers {
+		conn, err := h.tryConnect(addr)
+		if err != nil {
+			h.log.WithFields(
+				xlog.NewLogField("address", addr),
+				xlog.NewLogErrField(err),
+			).Error("error connecting to server")
+		}
+		go h.forwardToConn(addr, conn)
+	}
+
+	return h, nil
+}
+
+func (h *forwardHandler) Handle(buffer msgpack.Buffer) error {
+	// NB(xichen): the buffer contains newly flushed data so it's preferrable to keep
+	// it and drop the oldest buffer in queue in case the queue is full.
+	return h.enqueue(buffer, dropOldestInQueue)
+}
+
+func (h *forwardHandler) Close() {
+	h.Lock()
+	if h.closed {
+		h.Unlock()
+		return
+	}
+	h.closed = true
+	close(h.bufCh)
+	h.Unlock()
+
+	// Wait for the queue to be drained.
+	h.wg.Wait()
+}
+
+func (h *forwardHandler) tryConnect(addr string) (*net.TCPConn, error) {
+	conn, err := h.dialWithTimeoutFn(tcpProtocol, addr, h.connectTimeout)
+	if err != nil {
+		return nil, err
+	}
+	tcpConn := conn.(*net.TCPConn)
+	tcpConn.SetKeepAlive(h.connectionKeepAlive)
+	return tcpConn, nil
+}
+
+func (h *forwardHandler) enqueue(buffer msgpack.Buffer, dropType dropType) error {
+	h.RLock()
+	if h.closed {
+		h.RUnlock()
+		return errHandlerClosed
+	}
+	for {
+		select {
+		case h.bufCh <- buffer:
```

Comment:
Yes, `Handle` may be called from multiple goroutines. We store aggregated metrics with the same resolution in the same list, and when the list flushes, it calls `Handle` to flush the encoded data buffer when the size reaches MTU. 

The case you mentioned could happen, but only if (1) the queue is full and not draining fast enough, and (2) two goroutines flush at the same time (e.g., the 10s list and 1min list start flushing concurrently), and (3) it takes more time to perform one read from the channel than encoding a full buffer (otherwise the dequeuing goroutine should've successfully enqueued its buffer), which is likely not the case. 

So I'd say it's relatively rare. And even if it happens, as long as the dequeuing goroutine can enqueue the buffer in two or three iterations, it is not an issue, since the buffered data are already labeled with the correct timestamp.

Also, we use similar logic in shunts in prod and the contention among goroutines is much higher since any connection (including carbon metrics from the shunt server each host and the statsite boxes) can enqueue metrics.